File: core/src/main/java/io/github/msameer0/rhythmicrush/game/WorldPoolManager.java
package io.github.msameer0.rhythmicrush.game;

import com.badlogic.gdx.utils.Array;

import io.github.msameer0.rhythmicrush.game.engine.ObjectPool;
import io.github.msameer0.rhythmicrush.game.gameplay.blocks.Block;
import io.github.msameer0.rhythmicrush.game.gameplay.blocks.Slope;
import io.github.msameer0.rhythmicrush.game.gameplay.hazards.AbstractHazard;
import io.github.msameer0.rhythmicrush.game.gameplay.hazards.HalfSpike;
import io.github.msameer0.rhythmicrush.game.gameplay.hazards.SawBlade;
import io.github.msameer0.rhythmicrush.game.gameplay.hazards.Spike;
import io.github.msameer0.rhythmicrush.game.gameplay.interactables.orbs.AbstractOrb;
import io.github.msameer0.rhythmicrush.game.gameplay.interactables.orbs.BlackOrb;
import io.github.msameer0.rhythmicrush.game.gameplay.interactables.orbs.BlueOrb;
import io.github.msameer0.rhythmicrush.game.gameplay.interactables.orbs.GreenOrb;
import io.github.msameer0.rhythmicrush.game.gameplay.interactables.orbs.PinkOrb;
import io.github.msameer0.rhythmicrush.game.gameplay.interactables.orbs.RedOrb;
import io.github.msameer0.rhythmicrush.game.gameplay.interactables.orbs.YellowOrb;
import io.github.msameer0.rhythmicrush.game.gameplay.interactables.portals.AbstractPortal;
import io.github.msameer0.rhythmicrush.game.gameplay.interactables.portals.CubePortal;
import io.github.msameer0.rhythmicrush.game.gameplay.interactables.portals.GravityPortal;
import io.github.msameer0.rhythmicrush.game.gameplay.interactables.portals.MiniPortal;
import io.github.msameer0.rhythmicrush.game.gameplay.interactables.portals.ShipPortal;
import io.github.msameer0.rhythmicrush.game.gameplay.players.AbstractPlayer;
import io.github.msameer0.rhythmicrush.game.gameplay.players.Cube;
import io.github.msameer0.rhythmicrush.game.gameplay.players.Ship;

/**
 * Owns and manages all {@link ObjectPool} instances used by {@link GameWorld}.
 *
 * <p>Centralises pool declarations, obtain helpers, and free-all cleanup so
 * that {@code GameWorld} itself can focus on gameplay logic rather than memory
 * management boilerplate.</p>
 */
public class WorldPoolManager {

    // ── Pools ─────────────────────────────────────────────────────────────────

    private final ObjectPool<Block> blockPool = new ObjectPool<Block>() {
        @Override protected Block create()       { return new Block(); }
        @Override protected void  reset(Block b) { b.reset(); }
    };
    private final ObjectPool<Slope> slopePool = new ObjectPool<Slope>() {
        @Override protected Slope create()       { return new Slope(); }
        @Override protected void  reset(Slope s) { s.reset(); }
    };
    private final ObjectPool<Spike>     spikePool     = new ObjectPool<Spike>()     { @Override protected Spike     create() { return new Spike();     } @Override protected void reset(Spike     s) {} };
    private final ObjectPool<HalfSpike> halfSpikePool = new ObjectPool<HalfSpike>() { @Override protected HalfSpike create() { return new HalfSpike(); } @Override protected void reset(HalfSpike s) {} };
    private final ObjectPool<SawBlade>  sawBladePool  = new ObjectPool<SawBlade>()  { @Override protected SawBlade  create() { return new SawBlade();  } @Override protected void reset(SawBlade  s) {} };

    private final ObjectPool<CubePortal>    cubePortalPool    = new ObjectPool<CubePortal>()    { @Override protected CubePortal    create() { return new CubePortal();    } @Override protected void reset(CubePortal    p) {} };
    private final ObjectPool<ShipPortal>    shipPortalPool    = new ObjectPool<ShipPortal>()    { @Override protected ShipPortal    create() { return new ShipPortal();    } @Override protected void reset(ShipPortal    p) {} };
    private final ObjectPool<GravityPortal> gravityPortalPool = new ObjectPool<GravityPortal>() { @Override protected GravityPortal create() { return new GravityPortal(); } @Override protected void reset(GravityPortal p) {} };
    private final ObjectPool<MiniPortal>    miniPortalPool    = new ObjectPool<MiniPortal>()    { @Override protected MiniPortal    create() { return new MiniPortal();    } @Override protected void reset(MiniPortal    p) {} };

    private final ObjectPool<Cube> cubePool = new ObjectPool<Cube>() { @Override protected Cube create() { return new Cube(); } @Override protected void reset(Cube c) {} };
    private final ObjectPool<Ship> shipPool = new ObjectPool<Ship>() { @Override protected Ship create() { return new Ship(); } @Override protected void reset(Ship s) {} };

    private final ObjectPool<YellowOrb> yellowOrbPool = new ObjectPool<YellowOrb>() { @Override protected YellowOrb create() { return new YellowOrb(); } @Override protected void reset(YellowOrb o) { o.reset(); } };
    private final ObjectPool<BlueOrb>   blueOrbPool   = new ObjectPool<BlueOrb>()   { @Override protected BlueOrb   create() { return new BlueOrb();   } @Override protected void reset(BlueOrb   o) { o.reset(); } };
    private final ObjectPool<PinkOrb>   pinkOrbPool   = new ObjectPool<PinkOrb>()   { @Override protected PinkOrb   create() { return new PinkOrb();   } @Override protected void reset(PinkOrb   o) { o.reset(); } };
    private final ObjectPool<RedOrb>    redOrbPool    = new ObjectPool<RedOrb>()    { @Override protected RedOrb    create() { return new RedOrb();    } @Override protected void reset(RedOrb    o) { o.reset(); } };
    private final ObjectPool<BlackOrb>  blackOrbPool  = new ObjectPool<BlackOrb>()  { @Override protected BlackOrb  create() { return new BlackOrb();  } @Override protected void reset(BlackOrb  o) { o.reset(); } };
    private final ObjectPool<GreenOrb>  greenOrbPool  = new ObjectPool<GreenOrb>()  { @Override protected GreenOrb  create() { return new GreenOrb();  } @Override protected void reset(GreenOrb  o) { o.reset(); } };

    // ── Obtain helpers ────────────────────────────────────────────────────────

    public Block      obtainBlock()     { return blockPool.obtain(); }
    public Slope      obtainSlope()     { return slopePool.obtain(); }
    public Spike      obtainSpike()     { return spikePool.obtain(); }
    public HalfSpike  obtainHalfSpike() { return halfSpikePool.obtain(); }
    public SawBlade   obtainSawBlade()  { return sawBladePool.obtain(); }

    public CubePortal    obtainCubePortal()    { return cubePortalPool.obtain(); }
    public ShipPortal    obtainShipPortal()    { return shipPortalPool.obtain(); }
    public GravityPortal obtainGravityPortal() { return gravityPortalPool.obtain(); }
    public MiniPortal    obtainMiniPortal()    { return miniPortalPool.obtain(); }

    public Cube obtainCube() { return cubePool.obtain(); }
    public Ship obtainShip() { return shipPool.obtain(); }

    public YellowOrb obtainYellowOrb() { return yellowOrbPool.obtain(); }
    public BlueOrb   obtainBlueOrb()   { return blueOrbPool.obtain(); }
    public PinkOrb   obtainPinkOrb()   { return pinkOrbPool.obtain(); }
    public RedOrb    obtainRedOrb()    { return redOrbPool.obtain(); }
    public BlackOrb  obtainBlackOrb()  { return blackOrbPool.obtain(); }
    public GreenOrb  obtainGreenOrb()  { return greenOrbPool.obtain(); }

    // ── Free helpers ──────────────────────────────────────────────────────────

    public void freePlayer(AbstractPlayer player) {
        if (player instanceof Cube) cubePool.free((Cube) player);
        else if (player instanceof Ship) shipPool.free((Ship) player);
    }

    public void freeBlock(Block b) {
        if (b instanceof Slope) slopePool.free((Slope) b);
        else blockPool.free(b);
    }

    public void freeHazard(AbstractHazard h) {
        if (h instanceof Spike)          spikePool.free((Spike) h);
        else if (h instanceof HalfSpike) halfSpikePool.free((HalfSpike) h);
        else if (h instanceof SawBlade)  sawBladePool.free((SawBlade) h);
    }

    public void freePortal(AbstractPortal p) {
        if (p instanceof CubePortal)         cubePortalPool.free((CubePortal) p);
        else if (p instanceof ShipPortal)    shipPortalPool.free((ShipPortal) p);
        else if (p instanceof GravityPortal) gravityPortalPool.free((GravityPortal) p);
        else if (p instanceof MiniPortal)    miniPortalPool.free((MiniPortal) p);
    }

    public void freeOrb(AbstractOrb o) {
        if (o instanceof YellowOrb) yellowOrbPool.free((YellowOrb) o);
        else if (o instanceof BlueOrb)   blueOrbPool.free((BlueOrb) o);
        else if (o instanceof PinkOrb)   pinkOrbPool.free((PinkOrb) o);
        else if (o instanceof RedOrb)    redOrbPool.free((RedOrb) o);
        else if (o instanceof BlackOrb)  blackOrbPool.free((BlackOrb) o);
        else if (o instanceof GreenOrb)  greenOrbPool.free((GreenOrb) o);
    }

    // ── Bulk free ─────────────────────────────────────────────────────────────

    /**
     * Frees all active entities back to their pools and clears the provided arrays.
     * Also resets the typed subset arrays for hazards and portals.
     *
     * @param blocks          Active block array (will be cleared).
     * @param hazards         Active hazard array (will be cleared).
     * @param portals         Active portal array (will be cleared).
     * @param orbs            Active orb array (will be cleared).
     * @param activeSpikes    Typed hazard subset (will be cleared).
     * @param activeHalfSpikes Typed hazard subset (will be cleared).
     * @param activeSawBlades Typed hazard subset (will be cleared).
     * @param activeCubePortals    Typed portal subset (will be cleared).
     * @param activeShipPortals    Typed portal subset (will be cleared).
     * @param activeGravityPortals Typed portal subset (will be cleared).
     * @param activeMiniPortals    Typed portal subset (will be cleared).
     * @param blockCull  Index from which active (non-culled) blocks start.
     * @param hazardCull Index from which active hazards start.
     * @param portalCull Index from which active portals start.
     * @param orbCull    Index from which active orbs start.
     */
    public void freeAll(
        Array<Block> blocks, Array<AbstractHazard> hazards,
        Array<AbstractPortal> portals, Array<AbstractOrb> orbs,
        Array<Spike> activeSpikes, Array<HalfSpike> activeHalfSpikes, Array<SawBlade> activeSawBlades,
        Array<CubePortal> activeCubePortals, Array<ShipPortal> activeShipPortals,
        Array<GravityPortal> activeGravityPortals, Array<MiniPortal> activeMiniPortals,
        int blockCull, int hazardCull, int portalCull, int orbCull
    ) {
        for (int i = blockCull;  i < blocks.size;  i++) freeBlock(blocks.get(i));
        blocks.clear();

        for (int i = hazardCull; i < hazards.size; i++) freeHazard(hazards.get(i));
        hazards.clear();
        activeSpikes.clear();
        activeHalfSpikes.clear();
        activeSawBlades.clear();

        for (int i = portalCull; i < portals.size; i++) freePortal(portals.get(i));
        portals.clear();
        activeCubePortals.clear();
        activeShipPortals.clear();
        activeGravityPortals.clear();
        activeMiniPortals.clear();

        for (int i = orbCull; i < orbs.size; i++) freeOrb(orbs.get(i));
        orbs.clear();
    }
}

File: core/src/main/kotlin/io/github/msameer0/rhythmicrush/game/WorldPoolManager.kt
package io.github.msameer0.rhythmicrush.game

import com.badlogic.gdx.utils.Array
import io.github.msameer0.rhythmicrush.game.engine.ObjectPool
import io.github.msameer0.rhythmicrush.game.gameplay.blocks.Block
import io.github.msameer0.rhythmicrush.game.gameplay.blocks.Slope
import io.github.msameer0.rhythmicrush.game.gameplay.hazards.AbstractHazard
import io.github.msameer0.rhythmicrush.game.gameplay.hazards.HalfSpike
import io.github.msameer0.rhythmicrush.game.gameplay.hazards.SawBlade
import io.github.msameer0.rhythmicrush.game.gameplay.hazards.Spike
import io.github.msameer0.rhythmicrush.game.gameplay.interactables.orbs.AbstractOrb
import io.github.msameer0.rhythmicrush.game.gameplay.interactables.orbs.BlackOrb
import io.github.msameer0.rhythmicrush.game.gameplay.interactables.orbs.BlueOrb
import io.github.msameer0.rhythmicrush.game.gameplay.interactables.orbs.GreenOrb
import io.github.msameer0.rhythmicrush.game.gameplay.interactables.orbs.PinkOrb
import io.github.msameer0.rhythmicrush.game.gameplay.interactables.orbs.RedOrb
import io.github.msameer0.rhythmicrush.game.gameplay.interactables.orbs.YellowOrb
import io.github.msameer0.rhythmicrush.game.gameplay.interactables.pads.AbstractPad
import io.github.msameer0.rhythmicrush.game.gameplay.interactables.pads.BlackPad
import io.github.msameer0.rhythmicrush.game.gameplay.interactables.pads.BluePad
import io.github.msameer0.rhythmicrush.game.gameplay.interactables.pads.GreenPad
import io.github.msameer0.rhythmicrush.game.gameplay.interactables.pads.PinkPad
import io.github.msameer0.rhythmicrush.game.gameplay.interactables.pads.RedPad
import io.github.msameer0.rhythmicrush.game.gameplay.interactables.pads.YellowPad
import io.github.msameer0.rhythmicrush.game.gameplay.interactables.portals.AbstractPortal
import io.github.msameer0.rhythmicrush.game.gameplay.interactables.portals.CubePortal
import io.github.msameer0.rhythmicrush.game.gameplay.interactables.portals.GravityPortal
import io.github.msameer0.rhythmicrush.game.gameplay.interactables.portals.MiniPortal
import io.github.msameer0.rhythmicrush.game.gameplay.interactables.portals.ShipPortal
import io.github.msameer0.rhythmicrush.game.gameplay.players.AbstractPlayer
import io.github.msameer0.rhythmicrush.game.gameplay.players.Cube
import io.github.msameer0.rhythmicrush.game.gameplay.players.Ship

/**
 * Manages object pooling for frequently created and destroyed game objects to optimize performance.
 */
class WorldPoolManager {

    private val blockPool = object : ObjectPool<Block>() {
        override fun create() = Block()
        override fun reset(obj: Block) = obj.reset()
    }
    private val slopePool = object : ObjectPool<Slope>() {
        override fun create() = Slope()
        override fun reset(obj: Slope) = obj.reset()
    }

    private val spikePool = object : ObjectPool<Spike>() {
        override fun create() = Spike()
        override fun reset(obj: Spike) = obj.reset()
    }
    private val halfSpikePool = object : ObjectPool<HalfSpike>() {
        override fun create() = HalfSpike()
        override fun reset(obj: HalfSpike) = obj.reset()
    }
    private val sawBladePool = object : ObjectPool<SawBlade>() {
        override fun create() = SawBlade()
        override fun reset(obj: SawBlade) = obj.reset()
    }

    private val cubePortalPool = object : ObjectPool<CubePortal>() {
        override fun create() = CubePortal()
        override fun reset(obj: CubePortal) = obj.reset()
    }
    private val shipPortalPool = object : ObjectPool<ShipPortal>() {
        override fun create() = ShipPortal()
        override fun reset(obj: ShipPortal) = obj.reset()
    }
    private val gravityPortalPool = object : ObjectPool<GravityPortal>() {
        override fun create() = GravityPortal()
        override fun reset(obj: GravityPortal) = obj.reset()
    }
    private val miniPortalPool = object : ObjectPool<MiniPortal>() {
        override fun create() = MiniPortal()
        override fun reset(obj: MiniPortal) = obj.reset()
    }

    private val cubePool = object : ObjectPool<Cube>() {
        override fun create() = Cube()
        override fun reset(obj: Cube) {}
    }
    private val shipPool = object : ObjectPool<Ship>() {
        override fun create() = Ship()
        override fun reset(obj: Ship) {}
    }

    private val yellowOrbPool = object : ObjectPool<YellowOrb>() {
        override fun create() = YellowOrb()
        override fun reset(obj: YellowOrb) = obj.reset()
    }
    private val blueOrbPool = object : ObjectPool<BlueOrb>() {
        override fun create() = BlueOrb()
        override fun reset(obj: BlueOrb) = obj.reset()
    }
    private val pinkOrbPool = object : ObjectPool<PinkOrb>() {
        override fun create() = PinkOrb()
        override fun reset(obj: PinkOrb) = obj.reset()
    }
    private val redOrbPool = object : ObjectPool<RedOrb>() {
        override fun create() = RedOrb()
        override fun reset(obj: RedOrb) = obj.reset()
    }
    private val blackOrbPool = object : ObjectPool<BlackOrb>() {
        override fun create() = BlackOrb()
        override fun reset(obj: BlackOrb) = obj.reset()
    }
    private val greenOrbPool = object : ObjectPool<GreenOrb>() {
        override fun create() = GreenOrb()
        override fun reset(obj: GreenOrb) = obj.reset()
    }

    private val yellowPadPool = object : ObjectPool<YellowPad>() {
        override fun create() = YellowPad()
        override fun reset(obj: YellowPad) = obj.reset()
    }
    private val bluePadPool = object : ObjectPool<BluePad>() {
        override fun create() = BluePad()
        override fun reset(obj: BluePad) = obj.reset()
    }
    private val pinkPadPool = object : ObjectPool<PinkPad>() {
        override fun create() = PinkPad()
        override fun reset(obj: PinkPad) = obj.reset()
    }
    private val redPadPool = object : ObjectPool<RedPad>() {
        override fun create() = RedPad()
        override fun reset(obj: RedPad) = obj.reset()
    }
    private val blackPadPool = object : ObjectPool<BlackPad>() {
        override fun create() = BlackPad()
        override fun reset(obj: BlackPad) = obj.reset()
    }
    private val greenPadPool = object : ObjectPool<GreenPad>() {
        override fun create() = GreenPad()
        override fun reset(obj: GreenPad) = obj.reset()
    }

    fun obtainBlock(): Block = blockPool.obtain()
    fun obtainSlope(): Slope = slopePool.obtain()
    fun obtainSpike(): Spike = spikePool.obtain()
    fun obtainHalfSpike(): HalfSpike = halfSpikePool.obtain()
    fun obtainSawBlade(): SawBlade = sawBladePool.obtain()

    fun obtainCubePortal(): CubePortal = cubePortalPool.obtain()
    fun obtainShipPortal(): ShipPortal = shipPortalPool.obtain()
    fun obtainGravityPortal(): GravityPortal = gravityPortalPool.obtain()
    fun obtainMiniPortal(): MiniPortal = miniPortalPool.obtain()

    fun obtainCube(): Cube = cubePool.obtain()
    fun obtainShip(): Ship = shipPool.obtain()

    fun obtainYellowOrb(): YellowOrb = yellowOrbPool.obtain()
    fun obtainBlueOrb(): BlueOrb = blueOrbPool.obtain()
    fun obtainPinkOrb(): PinkOrb = pinkOrbPool.obtain()
    fun obtainRedOrb(): RedOrb = redOrbPool.obtain()
    fun obtainBlackOrb(): BlackOrb = blackOrbPool.obtain()
    fun obtainGreenOrb(): GreenOrb = greenOrbPool.obtain()

    fun obtainYellowPad(): YellowPad = yellowPadPool.obtain()
    fun obtainBluePad(): BluePad = bluePadPool.obtain()
    fun obtainPinkPad(): PinkPad = pinkPadPool.obtain()
    fun obtainRedPad(): RedPad = redPadPool.obtain()
    fun obtainBlackPad(): BlackPad = blackPadPool.obtain()
    fun obtainGreenPad(): GreenPad = greenPadPool.obtain()

    fun freePlayer(player: AbstractPlayer) {
        when (player) {
            is Cube -> cubePool.free(player)
            is Ship -> shipPool.free(player)
        }
    }

    fun freeBlock(b: Block) {
        when (b) {
            is Slope -> slopePool.free(b)
            else -> blockPool.free(b)
        }
    }

    fun freeHazard(h: AbstractHazard) {
        when (h) {
            is Spike -> spikePool.free(h)
            is HalfSpike -> halfSpikePool.free(h)
            is SawBlade -> sawBladePool.free(h)
        }
    }

    fun freePortal(p: AbstractPortal) {
        when (p) {
            is CubePortal -> cubePortalPool.free(p)
            is ShipPortal -> shipPortalPool.free(p)
            is GravityPortal -> gravityPortalPool.free(p)
            is MiniPortal -> miniPortalPool.free(p)
        }
    }

    fun freeOrb(o: AbstractOrb) {
        when (o) {
            is YellowOrb -> yellowOrbPool.free(o)
            is BlueOrb -> blueOrbPool.free(o)
            is PinkOrb -> pinkOrbPool.free(o)
            is RedOrb -> redOrbPool.free(o)
            is BlackOrb -> blackOrbPool.free(o)
            is GreenOrb -> greenOrbPool.free(o)
        }
    }

    fun freePad(p: AbstractPad) {
        when (p) {
            is YellowPad -> yellowPadPool.free(p)
            is BluePad -> bluePadPool.free(p)
            is PinkPad -> pinkPadPool.free(p)
            is RedPad -> redPadPool.free(p)
            is BlackPad -> blackPadPool.free(p)
            is GreenPad -> greenPadPool.free(p)
        }
    }

    fun freeAll(
        blocks: Array<Block>,
        hazards: Array<AbstractHazard>,
        portals: Array<AbstractPortal>,
        orbs: Array<AbstractOrb>,
        pads: Array<AbstractPad>,
        blockCull: Int,
        hazardCull: Int,
        portalCull: Int,
        orbCull: Int,
        padCull: Int
    ) {
        for (i in blockCull until blocks.size) freeBlock(blocks.get(i))
        blocks.clear()

        for (i in hazardCull until hazards.size) freeHazard(hazards.get(i))
        hazards.clear()

        for (i in portalCull until portals.size) freePortal(portals.get(i))
        portals.clear()

        for (i in orbCull until orbs.size) freeOrb(orbs.get(i))
        orbs.clear()

        for (i in padCull until pads.size) freePad(pads.get(i))
        pads.clear()
    }
}
